python/pyspark/sql/dataframe.py

```diff
@@ -2530,7 +2531,9 @@ def replace(  # type: ignore[misc]
         to_replace: Union[
             "LiteralType", List["LiteralType"], Dict["LiteralType", "OptionalPrimitiveType"]
         ],
-        value: Optional[Union["OptionalPrimitiveType", List["OptionalPrimitiveType"]]] = _NoValue,
+        value: Optional[
+            Union["OptionalPrimitiveType", List["OptionalPrimitiveType"], _NoValueType]
+        ] = _NoValue,
```

Comment:
I don't think we want to leak this to the end user here, but let's see what others have to say about it.

FYI @ueshin @xinrong-databricks @itholic

Reply:
Should I change these with `ignore`?

Reply:
@zero323 Which part do you mean you don't want to leak?

Reply:
@ueshin  I meant `_NoValueType`, but I honestly don't know.

Reply:
OK, I gave it another look and any changes to `pyspark.sql` seem to unnecessary for the core task (in-lining of `pyspark/__init__.py`) ‒ the whole type checking pipeline  passes just fine when omit `pyspark.sql` changes.

So let's resolve the conflicts,  revert whatever is not strictly necessary, double check and get this merged.